### PR TITLE
[ATen][Native][Special] Hermite polynomial prematurely return NaN if n is high

### PR DESCRIPTION
**Pull request by Aidyn-A**
Hermite polynomials diverge to NaN at high orders due to numerical overflow. The proposal is to prematurely return NaN of it is known that at this value it will be NaN.

According to my short test
```Python
import torch
device = "cuda"
dtype = torch.float32

x = torch.linspace(-1000, 1000, 100000, device=device, dtype=dtype)

for n in range(1024):
    if torch.special.hermite_polynomial_h(x, n).isnan().sum().item() == x.shape[0]:
        print(f"hermite_polynomial_h: all outputs are nans! n = {n}")
        break

for n in range(1024):
    if torch.special.hermite_polynomial_he(x, n).isnan().sum().item() == x.shape[0]:
        print(f"hermite_polynomial_he: all outputs are nans! n = {n}")
        break
```

The output values become NaNs at these orders:
```
hermite_polynomial_h: all outputs are nans! n = 53, dtype=torch.float32
hermite_polynomial_he: all outputs are nans! n = 61, dtype=torch.float32
hermite_polynomial_h: all outputs are nans! n = 272, dtype=torch.float64
hermite_polynomial_he: all outputs are nans! n = 304, dtype=torch.float64
```

Surely, it makes sense to increase the limit as a safety margin.

cc @msaroufim @ptrblck @eqy @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mruberry @kshitij12345 @manuelcandales @SherlockNoMad @angelayi

### CODE REVIEW

File: aten/src/ATen/native/Math.h [hunk 1]
```diff
@@ -3054,6 +3054,14 @@ inline C10_HOST_DEVICE T hermite_polynomial_h_forward(T x, int64_t n) {
         return x + x;
     }
 
+    if (std::is_same_v<T, float> && n > 128) {
+        return std::numeric_limits<T>::quiet_NaN();
+    } else if (std::is_same_v<T, double> && n > 512) {
+        return std::numeric_limits<T>::quiet_NaN();
+    } else if (n > 1024) {
+        return std::numeric_limits<T>::quiet_NaN();
+    }
```
Comment (malfet) [inline aten/src/ATen/native/Math.h:3063]: All CUDA versions supports constexprs right? If so, can one define `template<typename T> constexpr getHermitianLimit() { return 1024};` with specializations for float and double?
Can we do something like
```suggestion
    if (n > getHermitianLimit<T>()) {
        return std::numeric_limits<T>::quiet_NaN();
    }
```

**Review by malfet — APPROVED** [review]
LGTM, but would be nice to avoid code dedup (and use constexprs) by defining a specialized `constexpr int64 getHermitianNLimit()`

**Comment by atalman** [comment]
Looks like OSS test failed internally: ```test/inductor:test_inductor_cuda - test_pointwise_hermite_polynomial_h_cuda (test_torchinductor.GPUTests)```

**Comment by jeffdaily** [comment]
Either this PR needs to combine with #141150 or you need to take care to revert/reland them both together.  Reverting this PR caused the unskipped ROCm tests to start failing again.

**Comment by Aidyn-A** [comment]
@atalman would it be possible to run the failed test in the CI?

**Comment by jeffdaily** [comment]
@atalman or @malfet  can we get confirmation that the internal tests are passing so we don't have to revert this one again?

**Review by eqy — APPROVED** [review]

**Pull request closed without merging**